llvm/utils/TableGen/CodeGenDAGPatterns.cpp

```diff
@@ -271,20 +271,72 @@ void TypeSetByHwMode::dump() const {
   dbgs() << *this << '\n';
 }
 
-bool TypeSetByHwMode::intersect(SetType &Out, const SetType &In) {
-  bool OutP = Out.count(MVT::iPTR), InP = In.count(MVT::iPTR);
-  // Complement of In.
-  auto CompIn = [&In](MVT T) -> bool { return !In.count(T); };
+namespace {
+enum class TypeCat {
```

Comment:
I will likely rewrite this to be less overkill with the templates. At one point I thought it was going to be a lot more complicated, but as it turns out the function really just needs an MVT and an MVT->bool.